### ISSUE TITLE
Rebooting Linux support broken by chef-client

### ISSUE THREAD

**Issue by dfduarte** [test-kitchen/test-kitchen#1218]
## Description

Kitchen is breaking as soon as the reboot command runs on the Cookbook over a Linux OS

## Kitchen Version

```ruby
Test Kitchen version 1.15.0
```

## ChefDK Version

If you are running test-kitchen via ChefDK, `chef --version` will provide additional relevent version details.

```ruby
Chef Development Kit Version: 1.2.22
chef-client version: 12.18.31
delivery version: master (0b746cafed65a9ea1a79de3cc546e7922de9187c)
berks version: 2017-04-18T19:17:15.309226 18121] 2017-04-18T19:17:15.309387 18121] 2017-04-18T19:17:15.309495 18121] 2017-04-18T19:17:15.309591 18121] 2017-04-18T19:17:15.345073 18121] 2017-04-18T19:17:15.345246 18121] 5.6.0
kitchen version: 1.15.0
```

## Ruby Version

If you are not using test-kitchen via ChefDK, please provide the output of ruby --version.

## Platform Version

Development plataform:
Mac OS Sierra

Remote test machine:
CentOS 7 base AWS Official HVM AMI.

## Replication Case

Any cookbook having a reboot resource set in any situation, like

```ruby
reboot 'Rebooting to use new kernel' do
  action :reboot_now
end
```

## Kitchen Output

(...)

```ruby
       - execute sed -i 's/^#compress/compress/g' /etc/logrotate.conf && echo 'set to compress' || echo 'compress' > /etc/logrotate.conf
         * execute[chsh -s /sbin/nologin root] action run
           - execute chsh -s /sbin/nologin root
         * reboot[Rebooting to use new kernel] action reboot_now[2017-04-18T22:11:30+00:00] WARN: Rebooting system immediately, requested by 'Rebooting to use new kernel'
       [2017-04-18T22:11:30+00:00] WARN: Rebooting system immediately, requested by 'Rebooting to use new kernel'
       
           
           Running handlers:
           Running handlers complete
           Chef Client finished, 8/15 resources updated in 14 seconds
       [2017-04-18T22:11:31+00:00] WARN: Rebooting server at a recipe's request. Details: {:delay_mins=>0, :reason=>"Reboot by Chef", :timestamp=>2017-04-18 22:11:30 +0000, :requested_by=>"Rebooting to use new kernel"}
       [2017-04-18T22:11:31+00:00] WARN: Rebooting server at a recipe's request. Details: {:delay_mins=>0, :reason=>"Reboot by Chef", :timestamp=>2017-04-18 22:11:30 +0000, :requested_by=>"Rebooting to use new kernel"}
           
           Running handlers:
       [2017-04-18T22:11:31+00:00] ERROR: Running exception handlers
       [2017-04-18T22:11:31+00:00] ERROR: Running exception handlers
           Running handlers complete
       [2017-04-18T22:11:31+00:00] ERROR: Exception handlers complete
       [2017-04-18T22:11:31+00:00] ERROR: Exception handlers complete
           Chef Client failed. 8 resources updated in 14 seconds
       [2017-04-18T22:11:31+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       [2017-04-18T22:11:31+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       [2017-04-18T22:11:31+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
       [2017-04-18T22:11:31+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
       [2017-04-18T22:11:31+00:00] FATAL: Chef::Exceptions::Reboot: Rebooting server at a recipe's request. Details: {:delay_mins=>0, :reason=>"Reboot by Chef", :timestamp=>2017-04-18 22:11:30 +0000, :requested_by=>"Rebooting to use new kernel"}
       [2017-04-18T22:11:31+00:00] FATAL: Chef::Exceptions::Reboot: Rebooting server at a recipe's request. Details: {:delay_mins=>0, :reason=>"Reboot by Chef", :timestamp=>2017-04-18 22:11:30 +0000, :requested_by=>"Rebooting to use new kernel"}
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Failed to complete #converge action: [closed stream] on default-centos-72
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```

## Kitchen Diagnose

https://pastebin.com/T6TQXeaA

## NOTE:

This repository is for core issues with the Test-Kitchen framework, so if the issue is with a plugin you may be referred to file the issue with the appropriate plugin.

**Comment by coderanger** [comment]
You need to specifically enable the reboot handling feature. https://discourse.chef.io/t/test-kitchen-1-10-0-released/8721 has details though the `exit_status: :enabled` is no longer needed if you're running Chef 13 inside the VM (you need the client_fork bit though).

**Comment by dfduarte** [comment]
Oh one more thing:

This is VERY similar to this one: https://github.com/test-kitchen/test-kitchen/issues/1062

**Comment by coderanger** [comment]
What does your `.kitchen.yml` look like?

**Comment by dfduarte** [comment]
```ruby
---
<%
  #--------------------------------------------------------------------------
  # the driver_plugin can be overridden with an environment variable:
  #   $ KITCHEN_DRIVER=ec2 kitchen test
  # if not specified, defaults are used...
  #   - kitchen_driver_ci if environment variable CI=true is present
  #   - kitchen_driver_local is used otherwise (which defaults to vagrant)
  #--------------------------------------------------------------------------

  kitchen_driver_ci = 'ec2'
  kitchen_driver_local = 'vagrant'
  kitchen_driver_default = kitchen_driver_local

  if ENV['KITCHEN_DRIVER']
    kitchen_driver = ENV['KITCHEN_DRIVER']
  elsif ENV['CI']=="true"
    kitchen_driver = kitchen_driver_ci
  else
    kitchen_driver = kitchen_driver_default
  end
  puts "-----> driver_plugin: #{kitchen_driver.to_s}"
%>

driver_plugin: <%= kitchen_driver %>
driver_config:
  require_chef_omnibus: true
  <% if kitchen_driver == 'ec2' %>
  aws_access_key_id:      <%= ENV['AWS_ACCESS_KEY_ID'] %>
  aws_secret_access_key:  <%= ENV['AWS_SECRET_ACCESS_KEY'] %>
  aws_ssh_key_id:         <%= ENV['AWS_SSH_KEY_ID']         || "3cShared" %>
  region:                 <%= ENV['AWS_REGION']             || "us-east-1" %>
  availability_zone:      <%= ENV['AWS_AVAILABILITY_ZONE']  || "us-east-1b" %>
  instance_type: "t2.medium"
  security_group_ids: ["sg-e7c18b9d"]
  subnet_id: "subnet-616eff4a"
  vpc_id: "vpc-69201b0c"
  associate_public_ip: false
  <% end %>
  <% if kitchen_driver == 'vagrant' %>
  customize:
    memory: 2048
  <% end %>

<% if kitchen_driver == 'ec2' %>
transport:
  ssh_key: ~/workspace/Keys/Shared
  username: centos
<% end %>

provisioner:
  name: chef_zero
  # chef_omnibus_url: https://corp/raw/devops/ci-standards/master/test-kitchen/chef/install-chef.sh
  retry_on_exit_code:
    - 35
    - 20
    - 1
  wait_for_retry: 20
  max_retries: 1
  client_rb:
    exit_status: :enabled # Opt-in to the standardized exit codes
    client_fork: false  # Forked instances don't return the real exit code

platforms:
- name: centos-7.2
  <% if kitchen_driver == 'ec2' %>
  driver_config:
    image_id: ami-6d1c2007
  <% end %>

suites:
- name: default
  run_list:
    - recipe[centos_base_ami::default]
  attributes:
```

Note, I ommited some confidencial infos, but they are just some tags and the whole cookbook name, and won't affect the understanding. @coderanger

**Comment by dfduarte** [comment]
I'm trying to use the client_fork now, let see if it works...

**Comment by coderanger** [comment]
Check which version of Chef is being used, it will display at the top of the run.

**Comment by dfduarte** [comment]
@coderanger 
chef-13.0.118-1.el7

**Comment by coderanger** [comment]
So yeah, you can remove the `exit_status: :enabled` since that is now always on in Chef 13.

**Comment by dfduarte** [comment]
Well, no dice man =/

```
     * execute[chsh -s /sbin/nologin root] action run
           - execute chsh -s /sbin/nologin root
         * reboot[Rebooting to use new kernel] action reboot_now[2017-04-18T22:55:19+00:00] WARN: Rebooting system immediately, requested by 'Rebooting to use new kernel'
       [2017-04-18T22:55:19+00:00] WARN: Rebooting system immediately, requested by 'Rebooting to use new kernel'
       
           
           Running handlers:
           Running handlers complete
           Chef Client finished, 14/15 resources updated in 01 minutes 13 seconds
       [2017-04-18T22:55:19+00:00] WARN: Rebooting server at a recipe's request. Details: {:delay_mins=>0, :reason=>"Reboot by Chef", :timestamp=>2017-04-18 22:55:19 +0000, :requested_by=>"Rebooting to use new kernel"}
       [2017-04-18T22:55:19+00:00] WARN: Rebooting server at a recipe's request. Details: {:delay_mins=>0, :reason=>"Reboot by Chef", :timestamp=>2017-04-18 22:55:19 +0000, :requested_by=>"Rebooting to use new kernel"}
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Failed to complete #converge action: [closed stream] on default-centos-72
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```

**Comment by coderanger** [comment]
Hmm, the closed stream is odd, almost like it's rebooting before TK has a chance to react to the run ending? /cc @smurawski for a second opinion

**Comment by smurawski** [comment]
It is, the reboot command execution on linux runs synchronously and blocks until the reboot happens (see https://github.com/chef/chef/issues/5026), so Chef doesn't exit cleanly.  Probably shouldn't be to hard of a thing to fix, but there hasn't been a lot of clamoring for it to get done.  Reboot desires were mostly the hopes and dreams of those converging windows nodes. :)

**Comment by daxgames** [comment]
Having similar issues on Centos 7.3 vagrant box with Chef 12.15.19. I have a need to update the kernel and reboot before continuing and am doing :reboot_now

Relevant .kitchen.yml bits:

```
provisioner:
  name: chef_zero
  require_chef_omnibus: 12.15
  retry_on_exit_code:
  - 35
  max_retries: 1
  retry_on_exit: 1
  wait_for_retry: 40
  client_rb:
    exit_status: :enabled
    client_fork: false
```
 
Getting: 

```
           Chef Client finished, 21/21 resources updated in 03 minutes 08 seconds

           Running handlers:
           Running handlers complete
           Chef Client failed. 21 resources updated in 03 minutes 08 seconds
D      Cleaning up local sandbox in /var/folders/kt/44g108_17y36j17g80w3yk9w0000gn/T/build-centos-73-x86-64-sandbox-20170421-98778-1noi0y2
D      [SSH] shutting previous connection vagrant@127.0.0.1<{:user_known_hosts_file=>"/dev/null", :paranoid=>false, :port=>"2222", :compressi
on=>false, :compression_level=>0, :keepalive=>true, :keepalive_interval=>60, :timeout=>15, :keys_only=>true, :keys=>["/Users/daxgames/repos/g
e_github/corpwas/chef_repo/cookbooks/jenkinsdocker2/.kitchen/kitchen-vagrant/kitchen-jenkinsdocker2-build-centos-73-x86-64/.vagrant/m
achines/default/virtualbox/private_key"], :auth_methods=>["publickey"], :user=>"vagrant"}>
D      [SSH] closing connection to vagrant@127.0.0.1<{:user_known_hosts_file=>"/dev/null", :paranoid=>false, :port=>"2222", :compression=>fal
se, :compression_level=>0, :keepalive=>true, :keepalive_interval=>60, :timeout=>15, :keys_only=>true, :keys=>["/Users/daxgames/repos/githu
b/corpwas/chef_repo/cookbooks/jenkinsdocker2/.kitchen/kitchen-vagrant/kitchen-jenkinsdocker2-build-centos-73-x86-64/.vagrant/machines
/default/virtualbox/private_key"], :auth_methods=>["publickey"], :user=>"vagrant"}>

/opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/ruby_compat.rb:25:in `select': closed stream (IOError)
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/ruby_compat.rb:25:in `io_select'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/transport/packet_stream.rb:75:in `available_for_read?'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/transport/packet_stream.rb:87:in `next_packet'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/transport/session.rb:193:in `block in poll_message'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/transport/session.rb:188:in `loop'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/transport/session.rb:188:in `poll_message'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/connection/session.rb:474:in `dispatch_incoming_packets'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/connection/session.rb:225:in `preprocess'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/connection/session.rb:206:in `process'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/connection/session.rb:170:in `block in loop'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/connection/session.rb:170:in `loop'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/connection/session.rb:170:in `loop'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/net-ssh-3.2.0/lib/net/ssh/connection/session.rb:119:in `close'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/test-kitchen-1.15.0/lib/kitchen/transport/ssh.rb:115:in `close'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/test-kitchen-1.15.0/lib/kitchen/transport/ssh.rb:98:in `cleanup!'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/test-kitchen-1.15.0/lib/kitchen/instance.rb:285:in `cleanup!'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/test-kitchen-1.15.0/lib/kitchen/command.rb:205:in `run_action_in_thread'
        from /opt/chefdk/embedded/lib/ruby/gems/2.3.0/gems/test-kitchen-1.15.0/lib/kitchen/command.rb:169:in `block (2 levels) in run_action'
```

**Comment by smurawski** [comment]
@daxgames Same answer - that behavior is blocked on linux by chef/chef#5026

**Comment by daxgames** [comment]
@smurawski figured as much, just letting you know @dfduarte is not alone and even Linux Chef users need to reboot occaisionally.  :-)

**Comment by aaronlippold** [comment]
Given the blocking issues is closed and merged, what, if anything, needs to happen here to get this working?

**Comment by coderanger** [comment]
In theory, nothing. If the run ends with the correct exit code, the TK retry support will kick in on Linux exactly as it does on Windows.

**Comment by s-bernard** [comment]
I have the same issue even with latest chef which includes the fix.

**Comment by coderanger** [comment]
Make sure you use a delayed reboot, with enough time for Chef to fully exit before the restart.

**Comment by s-bernard** [comment]
I succeed to make it work:
- I had to put a delay_mins 1 to the reboot resource to give Chef enough time
- I increased ssh transport retries and retry_sleep

```yaml
transport:
  name: ssh
  connection_retries: 12
  connection_retry_sleep: 10

provisioner:
  name: chef_zero
  always_update_cookbooks: true
  retry_on_exit_code:
    - 35 # exit code signaling that the node is rebooting
  client_rb:
    client_fork: false # forked instances don't return the real exit code
```

You can the full code at https://gitlab.com/chef-platform/custom-kernel

**Comment by aaronlippold** [comment]
Odd. Vagrant / VirtualBox is still giving me odd behavior but I am sure that is my shaky chef knowledge.. I will keep poking and see if I can get it working. i.e. sometimes reboots, sometimes not.

**Comment by smurawski** [comment]
@s-bernard you should be able to just tweak the `wait_for_retry` setting in the provisioner rather than the connection retries for the transport.  That'll just add a bit of delay before retrying the transport connection.

**Comment by s-bernard** [comment]
@smurawski yes right but it's not necessarily better as it's a single big delay instead of multiple tries.

**Comment by cheeseplus** [comment]
This is no longer an issue but any further discussion/tips can be catalogued https://github.com/test-kitchen/kitchen-docs/issues/65
